File: bun/src/services/realtime/notification.processor.ts
import { NotificationType } from './notification'
import { INotificationExecutor } from './executor/notification.executor'

export interface Notification<T> {
  id: string
  message: T
}

export interface INotificationProcessor {
  notify<T>(type: NotificationType, notification: Notification<T>): Promise<void>
}

export class NotificationProcessor implements INotificationProcessor {
  notificationExecutors: INotificationExecutor[]

  constructor(notificationExecutors: INotificationExecutor[]) {
    this.notificationExecutors = notificationExecutors
  }

  async notify<T>(type: NotificationType, notification: Notification<T>): Promise<void> {
    const executions = this.notificationExecutors.map((executor) => executor.execute(type, notification))
    await Promise.all(executions)
  }
}

File: go/internal/processors/notification/notification_processor.go
package processor

import executors "github.com/marlpativ/healthz-curler/internal/executors/notification"

type NotificationProcessor interface {
	notify(notificationType string, notification interface{}) error
}

type notificationProcessor struct {
	notificationExecutors []executors.NotificationExecutor
}

func NewNotificationProcessor(notificationExecutors []executors.NotificationExecutor) NotificationProcessor {
	return &notificationProcessor{notificationExecutors: notificationExecutors}
}

func (n *notificationProcessor) notify(notificationType string, notification interface{}) error {
	for _, executor := range n.notificationExecutors {
		executor.Execute(notificationType, notification)
	}
	return nil
}
